Source/WebCore/css/values/primitives/CSSURL.cpp

```diff
@@ -0,0 +1,95 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "CSSURL.h"
+
+#include "CSSMarkup.h"
+#include "CSSParserContext.h"
+#include "CSSSerializationContext.h"
+#include "Document.h"
+
+namespace WebCore {
+namespace CSS {
+
+// MARK: - Serialization
+
+void Serialize<URL>::operator()(StringBuilder& builder, const SerializationContext& context, const URL& value)
+{
+    // https://drafts.csswg.org/cssom/#serialize-a-url
+
+    builder.append(nameLiteralForSerialization(CSSValueUrl), '(');
+
+    if (!value.resolved.isNull()) {
+        if (auto replacementURLString = context.replacementURLStrings.get(value.resolved.string()); !replacementURLString.isEmpty())
+            serializeString(replacementURLString, builder);
+        else if (context.shouldUseResolvedURLInCSSText)
+            serializeString(value.resolved.string(), builder);
+        else
+            serializeString(value.specified, builder);
+    } else
+        serializeString(value.specified, builder);
+
+    builder.append(')');
+}
+
+// MARK: Operations
+
+static URL completeURL(const String& string, const WTF::URL& baseURL)
+{
+    if (string.isEmpty() || string.startsWith('#'))
+        return URL { .specified = string, .resolved = WTF::URL { string } };
+    if (baseURL.isNull())
+        return URL { .specified = string, .resolved = WTF::URL { } };
+    return URL { .specified = string, .resolved = WTF::URL { baseURL, string } };
+}
+
+std::optional<URL> completeURL(const String& string, const CSSParserContext& context)
+{
+    if (string.isNull())
+        return { };
+
+    auto result = completeURL(string, context.baseURL);
+    if (context.mode == WebVTTMode && !result.resolved.protocolIsData())
+        return { };
+
+    return result;
+}
+
+std::optional<URL> completeURL(const String& string, const Document& document)
+{
+    if (string.isNull())
+        return { };
+    return completeURL(string, document.baseURL());
+}
+
+bool mayDependOnBaseURL(const URL& url)
+{
+    return !url.specified.isEmpty()
+        && !url.specified.startsWith('#')
+        && !protocolIs(url.specified, "data"_s);
```

Comment:
Doing scheme comparisons on a string won't work well. This won't account for inputs where the scheme contains `\t` or `\n`. It also seems very weird to special case `data:` URLs here. Is that really per specification?

Reply:
Not as an explanation, but this part is existing and is just moved. 

I believe this function is trying to just be conservative, hence the “may” in mayDependOnBaseURL, but I don’t know if that makes it ok. 

That said, we can use the resolved value here and check the URL protocol instead.

Reply:
Yeah I think the original code used `WebCoreURLInstance.protocolIsData()` when I skimmed through your patch. Doing that here as well would be preferable and maybe we should add a FIXME as this is deeply suspect.

Reply:
Should we rename the non-instance protocolIs to something more scary so people don’t make this mistake?

Reply:
Yeah that would be good. No great suggestion other than giving them an `Unsafe` suffix.

Reply:
Hm, actually, looking at the code for protocolIs(), which just calls protocolIsInternal(), it does look like it skips tabs and newlines -> https://github.com/WebKit/WebKit/blob/e7b2eb32d8a207ffb20c38d3513c4948dd5c3ae4/Source/WTF/wtf/URL.cpp#L869

Reply:
Oh sorry, I must have misremembered. (The original concern about the logic of this stands, but since this is a refactoring that's not really applicable.)